proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,111 @@
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/ezFC1TLGozU)
+
+## Motivation
+
+Currently there are at least two libraries to provide fault tolerance. It is best to uniform the technologies and define a standard so that micro service applications can adopt and the implementation of fault tolerance can be provided by the containers if possible.
+
+## Proposed solution
+
+Separate the responsibility of executing logic (Runnables/Callables/etc) from guiding when execution should take place (through retry policies, bulkheads, circuit breakers). In this way, failure handling strategies become configuration that can influence executions, and the execution API itself is just responsible for receiving some configuration and performing executions.
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* A Java API to privide RetryPolicy
+* A Java API to provide FallBack
+* A Java API to provide BulkHead
+* A Java API to provide CircuitBreaker
+
+## Detailed design
+
+The Design is built on CDI programming model. An instance of `Execution`, `RetryPolicy`, `CircuitBreaker` and `BulkHeader` can be injected to the clients and then they can be configured as follows.
+
+### RetryPolicy: A policy define the retry criteria
+
+
+```
+RetryPolicy rp = retryPolicy.retryOn(TimeOutException.class)
+  .withDelay(2, TimeUnit.SECONDS)
+  .withMaxRetries(2);
+```
+
+When `TimeOutException` was received, delay 2 seconds and then retry 2 more times.
+
+```
+Connection connection = execution.with(rp).withFallBack(null).get(this::connect)
+```
+
+The above suppress the `TimeOutException` and provide a default result.
+
+### CircuitBreaker: a rule to define when to close the circuit
+
+```
+CircuitBreaker cb = circuitBreaker
+  .withFailureThreshold(3, 10)
+  .withSuccessThreshold(5)
+  .withDelay(1, TimeUnit.MINUTES);
+Connection connect = execution.with(cb).run(this::connect);  
+```
+
+When 3 of 10 execution failures occurs on a circuit breaker, the circuit is opened and further execution requests fail with `CircuitBreakerOpenException`. After a delay of 1 min, the circuit is half-opened and trail executions are attempted to determine whether the circuit should be closed or opened again. If the trial executions exceed a success threshold 5, the breaker is closed again and executions will proceed as normal.
+
+### BulkHead
+
+```
+BulkHead bh = bulkHead.withPool("myPool");
```

Comment:
The instantiation of Bulkhead isn't addressed which should be fixed, and the bulkhead API still needs the most thought and input. 

Since bulkheads can/should be shared across executions, we may need a static way of identifying or looking up bulkhead instances. This could be facilitated by the API or with the help of a DI container as in your example.